lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,375 @@
+
+      def commit(method, endpoint, parameters=nil)
+        response = api_request(method, endpoint, parameters)
+        success = !response.key?("errors")
+        Response.new(success,
+          message_from(success, response),
+          response,
+          authorization: authorization_from(response),
+          # Neither avs nor cvv match are not exposed in the api.
+          avs_result: nil,
+          cvv_result: nil,
+          test: test?,
+          error_code: success ? nil : error_code_from(response)
+        )
+      end
+
+      def headers
+        {
+          'Authorization' => "Bearer " + @bearer_token,
+          'User-Agent' => "Square/v2 ActiveMerchantBindings/#{ActiveMerchant::VERSION}",
+          'X-Square-Client-User-Agent' => user_agent,
+          'Accept' => 'application/json',
+          'Content-Type' => 'application/json',
+          # uncomment the below when generating request/response json for unit tests.
+          'Accept-Encoding' => ''
+        }
+      end
+
+      def add_idempotency_key(post, options)
+        post[:idempotency_key] = (options[:idempotency_key] || generate_unique_id).to_s
+      end
+
+      def message_from(success, response)
+        # e.g. {"errors":[{"category":"INVALID_REQUEST_ERROR","code":"VALUE_TOO_LOW","detail":"`amount_money.amount` must be greater than 100.","field":"amount_money.amount"}]}
+        success ? "Success" : response['errors'].first['detail']
+      end
+
+      def authorization_from(response)
+        if response['transaction'] 
+          # This will return the transaction level identifier, of which there is >= 1 nested tender id
+          # which you may need to look up depending on your use case (e.g. refunding). That is available
+          #  in the response.transaction.tenders array.
+          return response['transaction']['id']
+        end
+      end
+
+      def api_request(method, endpoint, parameters)
+        json_payload = JSON.generate(parameters) if parameters
+        begin
+          raw_response = ssl_request(method, self.live_url + endpoint, json_payload, headers)
+          response = JSON.parse(raw_response)
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response = response_error(raw_response)
+        rescue JSON::ParserError
+          response = json_error(raw_response)
+        end
+        response
+      end
+
+      def response_error(raw_response)
+        begin
+          JSON.parse(raw_response)          
+        rescue JSON::ParserError
+          json_error(raw_response)
+        end
+      end
+
+      def json_error(raw_response)
+        msg = 'Invalid non-parsable json data response from the Square API. Please contact squareup.com/help/us/en/contact?prefill=developer_api if you continue to receive this message.'
+        msg += "  (The raw response returned by the API was #{raw_response.inspect})"
+        {
+          "errors" => [{
+            "category":"API_ERROR",
```

Comment:
my guess would be that they'll prefer arrows to colons because of old rubby compat